Add JumpList, Improve removing/renaming profiles

WelpThatWorked opened the pull request:
### Added
- Jumplist, stores 5 most recent profiles for faster access.

### Changed
- Jumplist recent items must be stored manually, as an existing jumplist cannot be retrieved from the system between sessions. Thus this list of recent profiles is stored in the form of their GUIDs in the context.
- Tray icon context menu now loads the last used profile between sessions.
- Tray icon context menu updates to reflect deleted or renamed profiles.
- Profiles are deactivated if activated when deleted.
- Upper dashboard active profile textblock updates to reflect renamed profiles.

WelpThatWorked committed:
Add Jumplist

Add jumplist that updates on loading and saving of context.

WelpThatWorked committed:
Merge branch 'develop' into feature/jumplist

WelpThatWorked committed:
Change to recent items

Change from using custom category to just recording recent profiles

WelpThatWorked committed:
Update CHANGELOG.md

WelpThatWorked committed:
Add workaround for JumpList

AddToRecentCategory seems to not respect the WorkingDirectory of the supplied JumpTask, so ensuring the right directory is set at startup is necessary.

WelpThatWorked committed:
Rework JumpList, Improve Systray icon

Redo JumpList into its own file, now handles profile renaming and removing. also handle profile renaming for SysTray icon.

WelpThatWorked committed:
Add behaviour for quitting without saving

WelpThatWorked committed:
Cleanup

WelpThatWorked committed:
Improve renaming and deleting behaviour

Disable profiles when deleted
Update trayicon and upper dashbar on renaming of active profile
Disable ActivateLastProfile on tray icon when last profile was deleted

WelpThatWorked committed:
Iteration #4

WelpThatWorked committed:
Fix first start-up crash

Forgot to initialize GUID list in case of context.xml not existing.

Snoothy commented on UCR.Core/Managers/SubscriptionsManager.cs:
```diff
@@ -45,7 +44,7 @@ public Profile GetActiveProfile()
         public bool ActivateLastProfile()
         {
-            return ActivateProfile(LastProfile);
+            return ActivateProfile(_context.Profiles.Find(p => p.Guid == _context.RecentProfiles[0]));
```
Implement this as a method in ProfilesManager

Snoothy commented on CHANGELOG.md:
```diff
@@ -12,6 +12,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - New setting: Minimize to tray on close
 - Added System Tray Icon
 - Added Hidden mode CLI Argument (-h|hidden)
+- Added JumpList for recent profiles
```
Also specify where this jump list is available from

Snoothy commented on UCR.Core/Context.cs:
```diff
@@ -202,6 +206,18 @@ public static T DeepXmlClone<T>(T obj)
 
         public void OnActiveProfileChangedEvent(Profile profile)
         {
+            if(profile != null)
+            {
+                if(!RecentProfiles.Any(p => p == profile.Guid))
+                {
+                    if (RecentProfiles.Count() == 5) RecentProfiles.RemoveAt(4);
+                }
+                else
+                {
+                    RecentProfiles.Remove(profile.Guid);
+                }
+                RecentProfiles.Insert(0, profile.Guid);
+            }
```
This should be implemented in ProfilesManager

Snoothy commented on UCR/Views/UCRTrayIcon.cs:
```diff
@@ -21,10 +21,19 @@ public UCRTrayIcon(MainWindow parent)
+            }
+            else
+            {
+                StartLastProfileStrip.Text = $"Activate Last Profile {parent.Context.Profiles.Find(p=>p.Guid==parent.Context.RecentProfiles[0]).Title}";
```
Use a ProfilesManager method instead of Search here

Snoothy commented on UCR/Views/UCRTrayIcon.cs:
```diff
@@ -101,6 +111,19 @@ private void TrayIcon_OnDoubleClick(object sender, EventArgs e)
+        {
+            if (_parent.Context.RecentProfiles.Count != 0)
+            {
+                StartLastProfileStrip.Text = $"Activate Last Profile: {_parent.Context.Profiles.Find(p => p.Guid == _parent.Context.RecentProfiles[0]).Title}";
```
Same as earlier, use ProfilesManager method to avoid duplicate code

Snoothy commented on UCR/Views/JumpList.cs:
```diff
@@ -0,0 +1,75 @@
+            oldJumpList = new System.Windows.Shell.JumpList();
+            foreach(Guid guid in _context.RecentProfiles)
+            {
+                Profile profile = _context.Profiles.Find(p => p.Guid == guid);
```
Use ProfilesManager method

WelpThatWorked committed:
Make specified improvements

- Clarify changelog
- Consolidate Guid->Profile to FindProfile function
- Move recent profile list updating to ProfilesManager

WelpThatWorked committed:
Close any ProfileWindows when minimized

Minor visual improvement, close any Profile Windows that might be open when the main window is minimized.
